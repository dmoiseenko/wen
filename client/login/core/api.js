import { postApiRequest } from '../../common/api';


export function login({ email, password }) { // eslint-disable-line import/prefer-default-export
  return postApiRequest(
    '/api/login',
    { email, password });
}
