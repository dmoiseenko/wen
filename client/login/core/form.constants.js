export const signInForm = 'signInForm';

export const emailField = 'emailField';

export const passwordField = 'passwordField';
