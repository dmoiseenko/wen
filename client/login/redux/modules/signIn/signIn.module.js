import { createAction } from 'redux-act';


export const signIn = createAction(
  'sign in'
);

export const signInSuccess = createAction(
  'sign in success'
);

export const signInFailure = createAction(
  'sign in failure'
);

export const reducer = {
  [signIn]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: true,
      loaded: false
    }
  }),

  [signInSuccess]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: false,
      loaded: true
    }
  }),

  [signInFailure]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: false,
      loaded: false
    }
  })
};
