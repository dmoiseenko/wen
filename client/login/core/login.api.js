import { postApiRequest } from '../../common/api';


export default function login({ email, password }) {
  return postApiRequest(
    '/api/login',
    { email, password });
}
