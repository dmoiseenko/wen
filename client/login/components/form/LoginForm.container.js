import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';

import LoginForm from './LoginForm';
import { signInForm } from '../../core/form.constants';
import { signIn } from '../../redux/modules/signIn/signIn.module';


export const mapDispatchToProps = {
  signIn
};

export const form = {
  form: signInForm
};

export default connect(null, mapDispatchToProps)(reduxForm(form)(LoginForm));
