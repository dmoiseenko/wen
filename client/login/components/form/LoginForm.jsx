import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';

import Input from '../../../common/components/input/Input';
import { emailField, passwordField } from '../../core/form.constants';


export default function LoginForm({ signIn }) {
  function handleClick(event) {
    event.preventDefault();
    signIn();
  }

  return (
    <form className="login-form">
      <Field
        name={emailField}
        component={Input}
        placeholder="mail@example.com"
        type="text"
        inputClassName="email-input"
        iconClassName="fa fa-user"
      />
      <Field
        name={passwordField}
        component={Input}
        placeholder="●●●●●●●"
        type="password"
        inputClassName="password-input"
        iconClassName="fa fa-lock"
      />
      <p className="control login">
        <button
          type="submit"
          onClick={handleClick}
          className="login-button"
        >
          Login
        </button>
      </p>
    </form>
  );
}

LoginForm.propTypes = {
  signIn: PropTypes.func.isRequired
};
