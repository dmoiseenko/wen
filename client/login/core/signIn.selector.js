import { formValueSelector } from 'redux-form';

import * as signInForm from './form.constants';


const signInFormSelector = formValueSelector(signInForm.signInForm);

export function password(state) {
  return signInFormSelector(state, signInForm.passwordField);
}

export function email(state) {
  return signInFormSelector(state, signInForm.emailField);
}
