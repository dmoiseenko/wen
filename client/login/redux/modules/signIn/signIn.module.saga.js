import { call, put, select, takeEvery } from 'redux-saga/effects';
import { startSubmit, stopSubmit } from 'redux-form';

import * as api from '../../../core/api';
import * as signInSelector from '../../../core/signIn.selector';
import * as signInForm from '../../../core/form.constants';
import * as location from '../../../../common/core/location';
import * as routes from '../../../../../common/routes';
import * as module from './signIn.module';


export function* signInSuccessSaga(response) {
  yield put.resolve(module.signInSuccess(response));
  yield put(stopSubmit(signInForm.signInForm));
  yield call(location.setLocation, routes.home);
}

export function* signInFailureSaga(error) {
  yield put.resolve(module.signInFailure(error.error));
  yield put(stopSubmit(signInForm.signInForm, {
    [signInForm.emailField]: error.error,
    [signInForm.passwordField]: error.error
  }));
}

export function* getEmailAndPasswordSaga() {
  const email = yield select(signInSelector.email);
  const password = yield select(signInSelector.password);

  return { email, password };
}

export function* signInSaga() {
  const { email, password } = yield call(getEmailAndPasswordSaga);

  yield put(startSubmit(signInForm.signInForm));

  const { response, error } = yield call(api.login, { email, password });

  if (response) {
    yield call(signInSuccessSaga, response);
  } else {
    yield call(signInFailureSaga, error);
  }
}

export function* watchSignInSaga() {
  yield takeEvery(module.signIn.getType(), signInSaga);
}
