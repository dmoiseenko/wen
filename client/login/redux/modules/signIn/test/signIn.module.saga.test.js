import { expectSaga, testSaga } from 'redux-saga-test-plan';
import { startSubmit, stopSubmit } from 'redux-form';

import {
  getEmailAndPasswordSaga,
  signInFailureSaga,
  signInSaga,
  signInSuccessSaga,
  watchSignInSaga
} from '../signIn.module.saga';
import * as module from '../signIn.module';
import * as signInForm from '../../../../core/form.constants';
import * as api from '../../../../core/api';
import * as location from '../../../../../common/core/location';
import * as routes from '../../../../../../common/routes';
import * as signInSelector from '../../../../core/signIn.selector';


describe('signInSaga', () => {
  const email = 'email';
  const password = 'password';
  const response = 'response';
  const error = 'error';

  it('should call signInSuccessSaga if login was successful', () => {
    testSaga(signInSaga)
      .next()
      .call(getEmailAndPasswordSaga)
      .next({ email, password })
      .put(startSubmit(signInForm.signInForm))
      .next()
      .call(api.login, { email, password })
      .next({ response })
      .call(signInSuccessSaga, response)
      .next()
      .isDone();
  });

  it('should call signInFailureSaga if login failed', () => {
    testSaga(signInSaga)
      .next()
      .call(getEmailAndPasswordSaga)
      .next({ email, password })
      .put(startSubmit(signInForm.signInForm))
      .next()
      .call(api.login, { email, password })
      .next({ error })
      .call(signInFailureSaga, error)
      .next()
      .isDone();
  });
});

describe('signInSuccessSaga', () => {
  const response = 'response';

  it('should have proper composition', () => {
    testSaga(signInSuccessSaga, response)
      .next()
      .put.resolve(module.signInSuccess(response))
      .next()
      .put(stopSubmit(signInForm.signInForm))
      .next()
      .call(location.setLocation, routes.home)
      .next()
      .isDone();
  });
});

describe('signInFailureSaga', () => {
  const error = { error: 'error' };

  it('should have proper composition', () => expectSaga(signInFailureSaga, error)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      }));
});

describe('signInSuccessSaga', () => {
  const response = 'response';

  it('should have proper composition', () => {
    testSaga(signInSuccessSaga, response)
      .next()
      .put.resolve(module.signInSuccess(response))
      .next()
      .put(stopSubmit(signInForm.signInForm))
      .next()
      .call(location.setLocation, routes.home)
      .next()
      .isDone();
  });
});

describe('getEmailAndPasswordSaga', () => {
  it('should have proper composition', () => {
    testSaga(getEmailAndPasswordSaga)
      .next()
      .select(signInSelector.email)
      .next()
      .select(signInSelector.password)
      .next()
      .isDone();
  });
});

describe('watchSignInSaga', () => {
  it('should have proper composition', () => {
    testSaga(watchSignInSaga)
      .next()
      .takeEveryEffect(module.signIn.getType(), signInSaga)
      .finish()
      .isDone();
  });
});
