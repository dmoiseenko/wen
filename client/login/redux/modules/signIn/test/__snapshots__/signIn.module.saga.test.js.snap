// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signInFailureSaga should have proper composition 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "error",
          "type": "[4] sign in failure",
        },
        "channel": null,
        "resolve": true,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "form": "signInForm",
          },
          "payload": Object {
            "emailField": "error",
            "passwordField": "error",
          },
          "type": "@@redux-form/STOP_SUBMIT",
        },
        "channel": null,
      },
    },
  ],
}
`;
